fedora/jellyfin.spec

```diff
@@ -6,16 +6,14 @@
 %global         dotnet_runtime  centos-x64
 %endif
 
-Name:           jellyfin
-Version:        10.5.0
+Name:           jellyfin-server
```

Comment:
Since you are changing the name of a package, you should add an:

```
Obsoletes: jellyfin < ${version}
```

so that the new package can upgrade the old one.

Reply:
There will be a metapackage eventually called "jellyfin", so not sure if this is required - let me know!

Reply:
You have to publish the metapackage and jellyfin-server at the same time then.  The metapackage will replace the existing `jellyfin` package and will `Require: jellyfin-server`.

Is there a reason that this specfile is not producing the metapackage and the `jellyfin-server` subpackage?  It might be more clear what is going on if that were the case.

Reply:
The idea is for the Metapackage to be built independently completely outside either this or the jellyfin-web repository, so there's no cross-dependency and both sides can be updated (and update the metapackage) independently. That said, I'm not sure if this is really going to work properly, but I figure it's worth a try.

Reply:
> The idea is for the Metapackage to be built independently completely outside either this or the jellyfin-web repository,

Why?  Metapackages are going to be dependent on the packaging system (RPM, debian, etc.) that they are built for.

What value do you see having it separate bringing?  I frankly see it as adding complication rather than bringing any kind of simplicity.

I ask this as a professional who does packaging of software for a living and who has been doing software packaging for more than 2 decades.

> so there's no cross-dependency

What kind of cross dependency are you trying to avoid?

From my professional perspective, the jellyfin backend package and the metapackage belong in the same packaging source.  The metapackage can then depend on some minimum (and/or maximum) version of the web-UI _addon_ to the server.  

> and both sides can be updated (and update the metapackage) independently.

I'm not even sure what goal that's trying to state.  Again, nothing about having the metapackage come from the main packaging source precludes any of this.

> That said, I'm not sure if this is really going to work properly, but I figure it's worth a try.

Seems like you are introducing more complication that is necessary, again from my professional perspective and experience.

Reply:
One of many reasons is that the web client moves much faster than the server so we'd like to release on that side more often. Another reason is that sometimes one will be ready for release and the other will have breaking bugs, so we are blocked when we could have otherwise tagged a release.

Reply:
> One of many reasons is that the web client moves much faster than the server so we'd like to release on that side more often.

OK.  That doesn't require a separate metapackage that needs to be updated.  One can have jellyfin-10.6.0 (metapackage) installed and continue to receive jellyfin-web updates as they are released using usual update commands such as `dnf upgrade`.  Getting the updated jellyfin-web does not require getting the updated metapackage.

> Another reason is that sometimes one will be ready for release and the other will have breaking bugs, so we are blocked when we could have otherwise tagged a release.

No you are not.  See the previous explanation.

Reply:
I only meant we are blocked when the two are bundled together in a single build process. It would definitely be ideal to avoid having to update whatever metapackage links the two, but I'm not extremely familiar with packaging systems. I think CentOS has some kind of labeling system for things like `epel-release` but I don't know if that's the method you were suggesting.

Reply:
We also have lots of instances where web changes have broken the build process, so separating the scripts into a web and server build will hopefully reduce this. The next step will be running the packaging in CI to check for issues before a release.

Reply:
> I only meant we are blocked when the two are bundled together in a single build process.

But the metapackage shouldn't ever need updating outside of the scope of a backend update also.  So there shouldn't be a case where you need to update the metapackage when the backend is not releasable.  Even if that very rare-to-never occasion ever did come up, you'd just do a point release from your most recent stable release, just updating the metapackge change.

> It would definitely be ideal to avoid having to update whatever metapackage links the two, 

I'm not really sure what you are trying to say here.

> but I'm not extremely familiar with packaging systems.

But I am.

> I think CentOS has some kind of labeling system for things like `epel-release` but I don't know if that's the method you were suggesting.

I don't know how that is relevant.

Basically the structure of the specfile would be

```
Name: jellyfin
Verison: 10.6.0
Release: 1
Requires: jellyfin-server = %{version}-%{release}
Requires: jellyfin-web < %{next_incompatible_API_version}

%description
The full Jellyfin experience complete with a Web-UI
...
%package server
%description server
Just the Jellyfin backend.  If you want this instance to provide a Web-UI also
please install the jellyfin package instead of this one.
...
```

With that, you can continue to make API-compatible jellyfin-web releases as long as you keep the major version of that package the same.  When you want to introduce a jellyfin-web that has breaking, non-backward compatible changes you bump it's major version.  The `Requires: jellyfin-web < %{next_incompatible_API_version}` is what prevents a given installation from installing too-new a jellyfin-web package.

But otherwise the above allows one to update one's jellyfin-web anytime one wants by simply using `dnf upgrade` (or `yum` or `apt-get`, etc.).

It's now on you (the _royal you_ -- the jellyfin team) to ensure that you don't release a jellyfin-web that breaks existing systems.  But that is nothing new and is a trade-off to the flexibility of this kind of package separation that you want to do -- for all of the benefits it provides.

From my professional experience, you probably want to introduce some regression testing to your jellyfin-web PR acceptance process that tests a given PR with the most recent released GA version of jellyfin-server to ensure you don't unknowingly introduce regressions.

This is a very important step in a development project when you decide to break a project up into micro-services, and my advise would be you are not ready to break it up until you have this process in place.  Or it's just a matter of time to when you release an mistakenly incompatible package.

Reply:
OK that makes sense to me.

My intent was to have the metapackage exist and be released only once with an arbitrarily high version (100 or so) so it would get updated once, and then never touch it again. Hence, having it separate. It would just always grab the "latest" version of both subpackages. Then new users install the `jellyfin` metapackage, and get whatever version of the subpackages, then the metapackage never gets touched again. That said, I too don't have much practical experience with packaging large server projects, though in my experience they always try to keep them together in one repo while we're trying to go the exact opposite way.

That said based on your explanation, I think this makes a lot of sense to me. I just don't really like it since it still ties releases of jellyfin-web to the main server repo in some way (i.e. having to update the metapackage and potentially need to trigger a release of jellyfin for a new build of jellyfin-web), which is what we want to avoid.

Reply:
So my intended flow would be something like:

*Existing user when 10.6.0 drops*
Upgrades old `jellyfin` to new metapackage `jellyfin`, gets `jellyfin-server` version 10.6.0 and `jellyfin-web` version 10.6.0 as part of the upgrade.

*New user*
Install `jellyfin`, gets `jellyfin-web` version 10.6.4 or w/e, gets `jellyfin-server` version 10.6.0 or w/e.

*New web release*
`jellyfin-web` gets updated to 10.6.5 without any changes to this repo at all.

Reply:
Just on that last point: yes I agree. We're breaking it up for 10.6.0 because we need to. We've been having nothing but problems trying to get these two repos to play nicely during the build stages since we first split it out. Too often we end up with a master web in a release or something to that effect because everything is so hacked together trying to support this half-split. This is the push for a full split. For now, the intent is to keep releasing them together, with the ability to do them separately in future releases when we have that infra in place.

Reply:
> My intent was to have the metapackage exist and be released only once with an arbitrarily high version (100 or so) so it would get updated once, and then never touch it again.

The same goal is achieved by just leaving it in the main packaging spec as I described above.  You only ever need touch it if you need to change some aspect of it, like restricting versions of the other packages (i.e. jellyfin-web) it wants to pull in.

> Hence, having it separate. It would just always grab the "latest" version of both subpackages.

Well, they are not subpackages if they are not in the same spec.  So just to be clear, let's not call packages built from other specfiles "subpackages".  And you don't always want the "latest" versions of packages.  You want the latest "compatible" versions.  You want the freedom to publish a jellyfin-web that is incompatible with older versions of jellyfin that people might still have installed without their installation unintentionally getting that newer jellyfin-web.

> Then new users install the `jellyfin` metapackage, and get whatever version of the subpackages, then the metapackage never gets touched again.

That's effectively how it works with true subpackages, where the metapackage and the subpackages (jellyfin-server in this case) exist in the same package.  That meta package, which is tied to the version of server can also specify a range of versions of other packages such as jellyfin-web.


> That said, I too don't have much practical experience with packaging large server projects, though in my experience they always try to keep them together in one repo

Yes.  To make things simpler.

> while we're trying to go the exact opposite way.

Yes.  You are making things more difficult.
 
> That said based on your explanation, I think this makes a lot of sense to me. I just don't really like it since it still ties releases of jellyfin-web to the main server repo in some way

Well, you have to "tie" at least to the degree of making sure that incompatible jellyfin-server and jellyfin-web packages are not installed together.  But that's why it's important to keep the metapackage in the same package as the jellyfin-server package.  That's what makes it easy to pair compatible versions.

> (i.e. having to update the metapackage and potentially need to trigger a release of jellyfin for a new build of jellyfin-web), which is what we want to avoid.

I'm not sure why I keep having to say this.  Clearly I am not being clear but I don't know how to say it and be more clear.  With a given version of the jellyfin/jellyfin-server package installed, you can freely upgrade jellyfin-web to any newer version (within it's compatibility range -- usually signified by the major version) **without** having to install a new metapackage.  I thought I provided a good example above in my example specfile outline.

To say it again, with a single given metapackage version installed that `Requires: jellyfin-web < 3`, one can update jellyfin-web-2.1 to 2.2 and then 2.3 and then 2.4, etc. as those versions are all released one after another simply doing `dnf upgrade`.  Again to be perfectly clear, that would just upgrade jellyfin-web and leave jellyfin/jellyfin-server alone (assuming there were no newer releases available).

It's only once you want to release jellyfin-web-3.0, where the bump in major version is to indicate that it's no longer backward compatible with jellyfin-server releases that were compatible with jellyfin-web-2.x that you need to update the metapackage associated with the server package to indicate the new API.  As you can see, this also prevent people who don't want to upgrade the server (and therefor the metapackage) from getting the incompatible jellyfin-web.

> So my intended flow would be something like:

Yes, this is the workflow that my proposal achieves.  I would suggest at this point a new versioning scheme for jellyfin-web however as releases of it and the server now become independent.

Reply:
OK I'm on board. Can you suggest exactly what changes need to be made to this specfile? I understand your blurb but not how to fit it in with what's there now.

Reply:
Something like this (completely untested):

```
%global         debug_package %{nil}
# Set the dotnet runtime
%if 0%{?fedora}
%global         dotnet_runtime  fedora-x64
%else
%global         dotnet_runtime  centos-x64
%endif

Name:           jellyfin
Version:        10.6.0
Release:        1%{?dist}
Summary:        The Free Software Media System Server backend and API
License:        GPLv3
URL:            https://jellyfin.media
# Jellyfin Server tarball created by `make -f .copr/Makefile srpm`, real URL ends with `v%{version}.tar.gz`
Source0:        jellyfin-server-%{version}.tar.gz
Source11:       jellyfin.service
Source12:       jellyfin.env
Source13:       jellyfin.sudoers
Source14:       restart.sh
Source15:       jellyfin.override.conf
Source16:       jellyfin-firewalld.xml

%{?systemd_requires}
BuildRequires:  systemd
BuildRequires:  libcurl-devel, fontconfig-devel, freetype-devel, openssl-devel, glibc-devel, libicu-devel
# Requirements not packaged in main repos
# COPR @dotnet-sig/dotnet or
# https://packages.microsoft.com/rhel/7/prod/
BuildRequires:  dotnet-runtime-3.1, dotnet-sdk-3.1
Requires: %{name}-server = %{version}-%{release}, %{name}-web >= 1, %{name}-web < 2
# Disable Automatic Dependency Processing
AutoReqProv:    no

%description
Jellyfin is a free software media system that puts you in control of managing and streaming your media.

%package server
# RPMfusion free
Requires(pre):  shadow-utils
Requires:       ffmpeg
Requires:       libcurl, fontconfig, freetype, openssl, glibc libicu

%description server
The Jellyfin media server backend.

%prep
%autosetup -n jellyfin-server-%{version} -b 0

%build

%install
export DOTNET_CLI_TELEMETRY_OPTOUT=1
export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
dotnet publish --configuration Release --output='%{buildroot}%{_libdir}/jellyfin' --self-contained --runtime %{dotnet_runtime} \
    "-p:GenerateDocumentationFile=false;DebugSymbols=false;DebugType=none" Jellyfin.Server
%{__install} -D -m 0644 LICENSE %{buildroot}%{_datadir}/licenses/jellyfin/LICENSE
%{__install} -D -m 0644 %{SOURCE15} %{buildroot}%{_sysconfdir}/systemd/system/jellyfin.service.d/override.conf
%{__install} -D -m 0644 Jellyfin.Server/Resources/Configuration/logging.json %{buildroot}%{_sysconfdir}/jellyfin/logging.json
%{__mkdir} -p %{buildroot}%{_bindir}
tee %{buildroot}%{_bindir}/jellyfin << EOF
#!/bin/sh
exec %{_libdir}/jellyfin/jellyfin \${@}
EOF
%{__mkdir} -p %{buildroot}%{_sharedstatedir}/jellyfin
%{__mkdir} -p %{buildroot}%{_sysconfdir}/jellyfin
%{__mkdir} -p %{buildroot}%{_var}/log/jellyfin
%{__mkdir} -p %{buildroot}%{_var}/cache/jellyfin

%{__install} -D -m 0644 %{SOURCE11} %{buildroot}%{_unitdir}/jellyfin.service
%{__install} -D -m 0644 %{SOURCE12} %{buildroot}%{_sysconfdir}/sysconfig/jellyfin
%{__install} -D -m 0600 %{SOURCE13} %{buildroot}%{_sysconfdir}/sudoers.d/jellyfin-sudoers
%{__install} -D -m 0755 %{SOURCE14} %{buildroot}%{_libexecdir}/jellyfin/restart.sh
%{__install} -D -m 0644 %{SOURCE16} %{buildroot}%{_prefix}/lib/firewalld/services/jellyfin.xml

%files server
%attr(755,root,root) %{_bindir}/jellyfin
%{_libdir}/jellyfin/*.json
%{_libdir}/jellyfin/*.dll
%{_libdir}/jellyfin/*.so
%{_libdir}/jellyfin/*.a
%{_libdir}/jellyfin/createdump
# Needs 755 else only root can run it since binary build by dotnet is 722
%attr(755,root,root) %{_libdir}/jellyfin/jellyfin
%{_libdir}/jellyfin/SOS_README.md
%{_unitdir}/jellyfin.service
%{_libexecdir}/jellyfin/restart.sh
%{_prefix}/lib/firewalld/services/jellyfin.xml
%attr(755,jellyfin,jellyfin) %dir %{_sysconfdir}/jellyfin
%config %{_sysconfdir}/sysconfig/jellyfin
%config(noreplace) %attr(600,root,root) %{_sysconfdir}/sudoers.d/jellyfin-sudoers
%config(noreplace) %{_sysconfdir}/systemd/system/jellyfin.service.d/override.conf
%config(noreplace) %attr(644,jellyfin,jellyfin) %{_sysconfdir}/jellyfin/logging.json
%attr(750,jellyfin,jellyfin) %dir %{_sharedstatedir}/jellyfin
%attr(-,jellyfin,jellyfin) %dir %{_var}/log/jellyfin
%attr(750,jellyfin,jellyfin) %dir %{_var}/cache/jellyfin
%{_datadir}/licenses/jellyfin/LICENSE

%pre server
getent group jellyfin >/dev/null || groupadd -r jellyfin
getent passwd jellyfin >/dev/null || \
    useradd -r -g jellyfin -d %{_sharedstatedir}/jellyfin -s /sbin/nologin \
    -c "Jellyfin default user" jellyfin
exit 0

%post server
# Move existing configuration cache and logs to their new locations and symlink them.
if [ $1 -gt 1 ] ; then
    service_state=$(systemctl is-active jellyfin.service)
    if [ "${service_state}" = "active" ]; then
        systemctl stop jellyfin.service
    fi
    if [ ! -L %{_sharedstatedir}/jellyfin/config ]; then
        mv %{_sharedstatedir}/jellyfin/config/* %{_sysconfdir}/jellyfin/
        rmdir %{_sharedstatedir}/jellyfin/config
        ln -sf %{_sysconfdir}/jellyfin  %{_sharedstatedir}/jellyfin/config
    fi
    if [ ! -L %{_sharedstatedir}/jellyfin/logs ]; then
        mv %{_sharedstatedir}/jellyfin/logs/* %{_var}/log/jellyfin
        rmdir %{_sharedstatedir}/jellyfin/logs
        ln -sf %{_var}/log/jellyfin  %{_sharedstatedir}/jellyfin/logs
    fi
    if [ ! -L %{_sharedstatedir}/jellyfin/cache ]; then
        mv %{_sharedstatedir}/jellyfin/cache/* %{_var}/cache/jellyfin
        rmdir %{_sharedstatedir}/jellyfin/cache
        ln -sf %{_var}/cache/jellyfin  %{_sharedstatedir}/jellyfin/cache
    fi
    if [ "${service_state}" = "active" ]; then
        systemctl start jellyfin.service
    fi
fi
%systemd_post jellyfin.service

%preun server
%systemd_preun jellyfin.service

%postun server
%systemd_postun_with_restart jellyfin.service

%changelog
...
```

I didn't make all of the changes in the example above, but with the above, since the main package %{name} is again _jellyfin_ you can revert all of the `%{name}` -> `jellyfin` changes where you were trying to preserve the _jellyfin_ name even though the package name had changed to _jellyfin-server_.  Basically all of the changes that look like:

```patch
-exec %{_libdir}/%{name}/%{name} \${@}
+exec %{_libdir}/jellyfin/jellyfin \${@}
 EOF
```

Also note the metapackage named _jellyfin_ has a:

```
Requires: %{name}-server = %{version}-%{release}, %{name}-web >= 1, %{name}-web < 2
```

This does two things:
1. Ensures that the server is updated with every update.
1. Allows the web package to be updated independently of the metapackage and server.  Obviously to be able to do this, those independent updates need to be backward compatible within their major version (1.x in my example).  But it also allows you to make compatibility breaking changes to the web package by updating it's major version (to 2.x for my example).

Reply:
That's fantastic, thanks so much @brianjmurrell 

I honestly like the idea of keeping most instances of "jellyfin" explicit throughout the spec more than using that interpolation everywhere. Mainly just to be really explicit about what each value is, aside from in that Requires line. As such I left those as-is while incorporating your new spec here. So far, it seems to work with minor adjustments.